src/main/java/seedu/awe/logic/commands/CalculatePaymentsCommand.java

```diff
@@ -82,32 +90,132 @@ public CommandResult execute(Model model) throws CommandException {
+            namesAndSurpluses.add(nameSurplusPair);
+        }
+        return namesAndSurpluses;
     }
 
     /**
      * Calculates payments to be made between Persons in the group.
      * @param group group for which payments are to be calculated
      * @return List of payments to make.
      */
-    public PriorityQueue<Payment> calculatePayments(Group group) {
-        // TODO: CALCULATES PAYMENTS
+    public String getPayments(Group group) throws CommandException {
+        List<Pair> namesAndSurplusesList = getNamesAndSurplusesList(group);
+        namesAndSurplusesList = sortPairs(namesAndSurplusesList);
+        List<Payment> payments = calculatePayments(namesAndSurplusesList);
+        String paymentsString = makePaymentsString(payments);
+        return paymentsString;
+    }
+
+    private boolean checkSumIsZero(List<Pair> pairs) {
+        double total = 0.00;
+        for (Pair pair: pairs) {
+            total += pair.getSurplus();
+        }
+        return total == 0.00;
+    }
+
+
+    private static Pair getSmallerPair(Pair p1, Pair p2) {
+        double p1AbsoluteSurplus = Math.abs(p1.getSurplus());
+        double p2AbsoluteSurplus = Math.abs(p2.getSurplus());
+        if (p1AbsoluteSurplus < p2AbsoluteSurplus) {
+            return p1;
+        } else if (p1AbsoluteSurplus > p2AbsoluteSurplus) {
+            return p2;
+        } else {
+            return null;
+        }
+    }
 
-        PriorityQueue<Payment> paymentsQueue = new PriorityQueue<>();
+    /**
+     * Tabulates payments to make between all persons to balance payments and expenses.
+     * @param pairs Name and Surplus/Deficit of Person.
+     * @return List of payments to balance payments and expenses.
+     * @throws CommandException thrown if a discrepancy in payments is observed.
+     */
+    public List<Payment> calculatePayments(List<Pair> pairs) throws CommandException {
+        if (!checkSumIsZero(pairs)) {
+            throw new CommandException("There appears to be a discrepancy within your payments.");
+        }
+        List<Payment> payments = new ArrayList<>();
+        while (!pairs.isEmpty()) {
+            if (pairs.size() == 1) {
+                throw new CommandException("There appears to be a discrepancy within your payments.");
+            }
+            Pair pairWithLowestSurplus = pairs.get(0);
+            Pair pairWithHighestSurplus = pairs.get(pairs.size() - 1);
+            Payment paymentToAdd = calculatePayment(pairWithLowestSurplus, pairWithHighestSurplus);
+            payments.add(paymentToAdd);
+            Pair smallerPair = getSmallerPair(pairWithLowestSurplus, pairWithHighestSurplus);
+            if (smallerPair.equals(pairWithHighestSurplus)) {
+                pairs.remove(pairs.size() - 1);
+            } else if (smallerPair.equals(pairWithLowestSurplus)) {
+                pairs.remove(0);
+            } else {
+                pairs.remove(0);
+                pairs.remove(pairs.size() - 1);
+            }
```

Comment:
Is there's a chance that this will go haywire, since smallerPair returns null if they are the same?